src/action/list_directory.rs

```diff
@@ -0,0 +1,553 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! A handler and associated types for the list directory action.
+//!
+//! A list directory action stats all files in the provided directory.
+
+use crate::session::{self, Session};
+use rrg_proto::{ListDirRequest, StatEntry};
+
+use ioctls;
+use std::fs::{self, File};
+use std::path::PathBuf;
+use std::os::raw::c_long;
+use std::os::unix::fs::MetadataExt;
+use std::os::unix::io::AsRawFd;
+use std::fmt::{Display, Formatter};
+
+#[derive(Debug)]
+enum Error {
+    ReadPath(std::io::Error),
+}
+
+impl std::error::Error for Error {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use Error::*;
+
+        match *self {
+            ReadPath(ref error) => Some(error),
+        }
+    }
+}
+
+impl Display for Error {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        use Error::*;
+
+        match *self {
+            ReadPath(ref error) => {
+                write!(fmt, "unable to read path: {}", error)
+            }
+        }
+    }
+}
+
+impl From<Error> for session::Error {
+
+    fn from(error: Error) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+
+/// A response type for the list directory action.
+pub struct Response {
+    st_mode: u64,
+    st_ino: u32,
+    st_dev: u32,
+    st_nlink: u32,
+    st_uid: u32,
+    st_gid: u32,
+    st_size: u64,
+    st_atime: u64,
+    st_mtime: u64,
+    st_ctime: u64,
+    st_blocks: u32,
+    st_blksize: u32,
+    st_rdev: u32,
+    st_flags_linux: u32,
+    symlink: Option<String>,
+    pathspec: PathSpec,
+}
+
+/// A request type for the list directory action.
+pub struct Request {
+    pathspec: PathSpec,
+}
+
+enum PathType {
+    OS,
+    TSK,
+    Registry,
+    TMPFile,
+    NTFS,
+}
+
+enum PathOption {
+    CaseInsensitive,
+    CaseLiteral,
+    Regex,
+    Recursive,
+}
+
+struct PathSpec {
+    path_options: Option<PathOption>,
+    pathtype: PathType,
+    path: PathBuf,
+}
+
+pub fn handle<S: Session>(session: &mut S, request: Request)
+                          -> session::Result<()> {
+    let dir_path = &request.pathspec.path;
+    // In case if Result<DirEntry> matches the error branch, we won't give
+    // any info about that file, but other info isn't effected:
+    let (dir_entries, _): (Vec<_>, Vec<_>) = dir_path.read_dir()
+        .map_err(Error::ReadPath)?.partition(Result::is_ok);
+    // Code won't panic because of unwrap(). We made the partition above
+    let mut paths: Vec<PathBuf> = dir_entries.into_iter().map(Result::unwrap)
+        .map(|entry| entry.path()).collect();
```

Comment:
(I cannot reply to the previous thread on this issue)

I am pretty sure you can get rid of these unwraps if you use [`filter_map`].

[`filter_map`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map

Reply:
Yep, thank you, done.